motor/frameworks/asyncio.py

```diff
@@ -198,59 +154,54 @@ def settimeout(self, timeout):
     @asyncio_motor_sock_method
     @asyncio.coroutine
     def connect(self):
-        protocol_factory = lambda: self
-
-        # TODO: will call getaddrinfo again.
-        host, port = self.options.address
-        self._transport, protocol = yield from self.loop.create_connection(
-            protocol_factory, host, port,
-            ssl=self.ctx)
+        is_unix_socket = (self.options.family == getattr(socket,
+                                                         'AF_UNIX', None))
+        if self.options.use_ssl:
+            # TODO: cache at Pool level.
+            ctx = ssl.SSLContext(ssl.PROTOCOL_SSLv23)
+            if self.options.certfile is not None:
+                ctx.load_cert_chain(self.options.certfile,
+                                    self.options.keyfile)
+            if self.options.ca_certs is not None:
+                ctx.load_verify_locations(self.options.ca_certs)
+            if self.options.cert_reqs is not None:
+                ctx.verify_mode = self.options.cert_reqs
+                if ctx.verify_mode in (ssl.CERT_OPTIONAL, ssl.CERT_REQUIRED):
+                    ctx.check_hostname = True
+        else:
+            ctx = None
+
+        if is_unix_socket:
+            path = self.options.address[0]
+            reader, writer = yield from asyncio.open_unix_connection(
+                path, loop=self.loop, ssl=ctx)
+        else:
+            host, port = self.options.address
+            reader, writer = yield from asyncio.open_connection(
+                host=host, port=port, ssl=ctx, loop=self.loop)
+            sock = writer.transport.get_extra_info('socket')
+            sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
+            sock.setsockopt(socket.SOL_SOCKET, socket.SO_KEEPALIVE,
+                            self.options.socket_keepalive)
+        self._reader, self._writer = reader, writer
 
     def sendall(self, data):
         assert greenlet.getcurrent().parent is not None,\
             "Should be on child greenlet"
 
         # TODO: backpressure? errors?
-        self._transport.write(data)
+        self._writer.write(data)
 
     @asyncio_motor_sock_method
     @asyncio.coroutine
     def recv(self, num_bytes):
-        while self._buffer_len < num_bytes:
-            yield from self._recv_future
-
-        data = bytes().join(self._buffer)
-        rv = data[:num_bytes]
-        remainder = data[num_bytes:]
-
-        self._buffer.clear()
-        if remainder:
-            self._buffer.append(remainder)
-
-        self._buffer_len = len(remainder)
-
+        rv = yield from self._reader.readexactly(num_bytes)
```

Comment:
Question: Is it possible to save a function call here somehow, since recv is just a decorated synonym for self._reader.readexactly?